bpo-43817: Add inspect.get_annotations().

larryhastings opened the pull request:
`inspect.get_annotations(o)` safely retrieves and return
 the annotations from `o`.  It handles un-stringizing
stringized annotations, it works around the quirks of
annotations on various objects, and it is super-safe.

`inspect.signature()` now calls `inspect.get_annotations()`
to get the annotations for an object.

<!-- issue-number: [bpo-43817](https://bugs.python.org/issue43817) -->
https://bugs.python.org/issue43817
<!-- /issue-number -->

larryhastings committed:
Add inspect.get_annotations().

inspect.get_annotations(o) safely retrieves and return
 the annotations from o.  It handles un-stringizing
stringized annotations, it works around the quirks of
annotations on various objects, and it is super-safe.

inspect.signature() now calls inspect.get_annotations()
to get the annotations for an object.

larryhastings committed:
Added news file.

larryhastings committed:
Fix whitespace in unimportant test file.

Thank goodness!

gvanrossum commented on Lib/inspect.py:
```diff
@@ -60,6 +60,133 @@
+        globals = obj_globals
+
+    return_value = {key:
+        value if not isinstance(value, str) else eval(value, globals, locals)
```
Drive-by comment: in get_type_hints() we recently introduced a feature (promised by PEP 563 but not previously implemented) where the locals are given a correct default for class annotations (GH-24201). Maybe we should do that here too?

larryhastings replied:
Good point!  I'll take a look, once I get a PR for bpo-43901 out the door.

JelleZijlstra commented on Lib/inspect.py:
```diff
@@ -2151,7 +2279,7 @@ def _signature_from_builtin(cls, func, skip_bound_arg=True):
 
 def _signature_from_function(cls, func, skip_bound_arg=True,
-                             globalns=None, localns=None):
+                             globals=None, locals=None, eval_str=ONLY_IF_STRINGIZED):
```
This may have backward compatibility problems: `inspect.signature` will start raising on 3.10 for a function that has all string annotations but uses names that aren't in the provided namespace. I feel like using `eval_str=False` here is safer.

larryhastings replied:
Before PEP 563 was reverted, inspect.signature() called typing.get_type_hints() to get the annotations for the function, which also called eval() on string annotations and could raise exceptions.  I was emulating this behavior.

Also, this is why the eval_str parameter exists.  Yes, it may start raising an exception for users in this situation, but it also allows for an easy fix.

There are good arguments to be made here for a default of False, and also for a default of ONLY_IF_STRINGIZED.  A default of ONLY_IF_STRINGIZED is bad for users who

- turn on stringized annotations,
- use annotations of values that aren't defined in their namespace, *and*
- examine the signature at runtime.

But it's better for literally everyone else, which is the vast majority of users.  So, as much as I don't like introducing a breaking change, I still think a default of ONLY_IF_STRINGIZED is the better choice.

isidentical replied:
> Before PEP 563 was reverted, inspect.signature() called typing.get_type_hints() to get the annotations for the function, which also called eval() on string annotations and could raise exceptions. I was emulating this behavior.

More specifically, the implementation was like this;

```py
def _get_type_hints(func):
    try:
        return typing.get_type_hints(func)
    except Exception:
        # First, try to use the get_type_hints to resolve
        # annotations. But for keeping the behavior intact
        # if there was a problem with that (like the namespace
        # can't resolve some annotation) continue to use
        # string annotations
        return func.__annotations__
```

larryhastings replied:
Huh.  I specifically experimenting with typing.get_type_hints() and getting it to raise a NameError.  Is that catchall except statement recent?

isidentical replied:
> Huh. I specifically experimenting with typing.get_type_hints() and getting it to raise a NameError. Is that catchall except statement recent?

No, it was there since the PEP 563 became default.

larryhastings replied:
Oh, okay.  So, what I'm emulating is the behavior of typing.get_type_hints() as it shipped in 3.9 (and presumably earlier versions).

    from __future__ import annotations

    def foo(a:int, b:flarg): pass

    import typing
    print(typing.get_type_hints(foo))

That raises a `NameError` in Python 3.9.4.

Fidget-Spinner replied:
FWIW, ``typing.get_type_hints`` also raises a ``NameError`` for that code example in the latest 3.10 from ``master``.

larryhastings replied:
Good.  I think that's the right choice for `typing.get_type_hints()`, and for `inspect.get_annotations()` and `inspect.signature()` too.

Note that, unlike `typing.get_type_hints()`, `inspect.get_annotations()` and `inspect.signature()` give the caller control over whether or not `eval()` is called.  If your annotations are stringized, and you only want to see the strings anyway, you can call them like so:

    ann = inspect.get_annotations(o, eval_str=False)
    sig = inspect.get_signature(o, eval_str=False)

larryhastings replied:
I've changed the default to eval_str=False.

JelleZijlstra commented on Lib/inspect.py:
```diff
@@ -3106,10 +3236,10 @@ def __str__(self):
 
 
-def signature(obj, *, follow_wrapped=True, globalns=None, localns=None):
+def signature(obj, *, follow_wrapped=True, globals=None, locals=None, eval_str=ONLY_IF_STRINGIZED):
```
Renaming the keyword arguments is definitely unsafe

larryhastings replied:
Why?  They are new in 3.10.

JelleZijlstra replied:
Sorry, didn't realize they are new.

larryhastings replied:
What happened was, those arguments were added to `inspect.signature()` for Python 3.10 as part of the "PEP 563 becoming default behavior" work.  `inspect.signature()` was changed to call  `typing.get_type_hints()`.  When that work was backed out, these changes were overlooked and left in.

I changed the names because I thought they should match the arguments to `eval()`.  If for some reason we decided that `inspect.signature()` shouldn't call `inspect.get_annotations()` after all, the arguments should simply be removed.

Fidget-Spinner commented on Misc/NEWS.d/next/Library/2021-04-22-04-12-13.bpo-43817.FQ-XlH.rst:
```diff
@@ -0,0 +1,6 @@
+stringized.
+
+Modified inspect.signature() to use get_annotations() to retrieve
+annotations.
```
```suggestion
Add ``inspect.get_annotations(o)``, which safely retrieves and returns the
annotations for ``o``.  Also un-stringizes the annotations if they are
stringized.

Modified ``inspect.signature()`` to use ``get_annotations()`` to retrieve
annotations.
```

larryhastings replied:
Thanks!  I'm not sure I'm sophisticated enough at Github to merge your suggestion using Github itself, but I'll make that change when I update the PR.

larryhastings committed:
Added ReST markup to news blurb.

Thanks, Fidget_Spinner!

larryhastings committed:
get_annotations() now provides locals for classes.

If an explicit locals dict is not passed in, get_annotations(cls)
will now use the cls namespace as locals by default.

larryhastings committed:
Minor doc touchups for inspect.signature().

larryhastings committed:
Changed eval_str to default to False.

inspect.get_annotations() and inspect.signature() will now
only call eval() on string values by explicit user request.
Calling eval() by default is too opinionated for the standard
library.

Neither function catches exceptions, so if the eval() raises
an exception, it'll unwind up to the caller.

Also, removed the special eval_str=ONLY_IF_STRINGIZED logic
entirely.  The user now tells us explicitly whether or not
to call eval() on strings--so we don't need the heuristic
for anything.

larryhastings committed:
Updated the news blurb to match the new semantics.

larryhastings committed:
Added tests for get_annotations call in signature.

inspect.signature() and inspect.Signature_from_callable()
both call inspect.get_annotations() to compute the annotations.
Both functions also pass through their globals, locals, and
eval_str arguments.  This commit adds some tests just to
ensure that the values are passed through correctly and everything
is working.  And who doesn't feel more secure with some extra tests!

larryhastings committed:
Added important newline to unimportant test module.

JelleZijlstra commented on Lib/inspect.py:
```diff
@@ -1898,7 +2015,7 @@ def _signature_is_functionlike(obj):
             (defaults is None or isinstance(defaults, tuple)) and
             (kwdefaults is None or isinstance(kwdefaults, dict)) and
-            isinstance(annotations, dict))
+            isinstance(annotations, (dict, None)))
```
`isinstance(obj, None)` doesn't work.

JelleZijlstra replied:
Interestingly it gets checked lazily, which is probably why this didn't get picked up by tests:
```
>>> isinstance(1, (dict,None))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: isinstance() arg 2 must be a type, a tuple of types or a union
>>> isinstance({}, (dict,None))
True
>>> 
```
I'm not sure when a function's `__annotations__` would be None.

larryhastings replied:
Oh!  That `isinstance` `None` stuff sure is dumb.  Thanks for spotting it!

Currently I don't know why someone would set `__annotations__`  to `None` either, but I note that it is specifically permitted by the function object.  It only allows setting `__annotations__` to a dict or `None`.  (Classes and modules are agnostic and will permit setting `__annotations__` to any value.)

It's probably too late for Python 3, but if there ever is a Python 4, I think `__annotations__` should enforce always being either a dict or `None` on all objects that support it, and that it should default to `None` when the object has no annotations.

larryhastings replied:
Also note just a few lines above:

    annotations = getattr(obj, '__annotations__', None)

If `__annotations__` isn't set, `annotations` will be `None`.  It's true that `__annotations__` is always set on CPython function objects, but that isn't guaranteed of other callables / objects that may be tested by this function.

larryhastings replied:
Oh, it's more complicated than I remembered.  You're allowed to set `func.__annotations__ = None`, but it internally just sets its annotations storage to `NULL`, and then if you retrieve annotations it lazy-creates a new dict.  So `func.__annotations__` on a real CPython function object will never ever be anything but a dict.  (Again, not so for classes, modules, wrapped functions, etc ad nauseum.)

And I just fixed this bug and refreshed the PR.  Marking this as "resolved".

larryhastings committed:
Fixed regression if __annotations__ is None.

__annotations__ is allowed to be None in some circumstances,
and there is explicit code to handle it in the inspect module.
But I introduced a bug ("isinstance(o, None)"), and this showed
the code path wasn't being executed anyway, so now there's a
bugfix and a test too.

larryhastings committed:
Rewrite news blurb.
